### PR TITLE
US132939: load server data

### PR DESCRIPTION
**Pull request by johngwilkinson**
I've matched the current server response for now, but that may change slightly. Hoping to get this foundational stuff in in the meantime.

Todo in another PR: error handling (currently sets a flag but doesn't use it)

Quality Notes
- functional
  - demo & e2e (server data is currently canned): graph appears as expected
- security/performance/devops/ux/docs: n/a
- lit2: no new shadowRoot references, I believe

FYI @Brightspace/draco-vis

### CODE REVIEW

File: model/dataApiClient.js [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+ * Wraps the BDP API adoption data routes and related LMS calls
+ */
+export class DataApiClient {
+	constructor(metronEndpoint) {
+		this._metronEndpoint = metronEndpoint;
+		this._locations = null;
+		this._launchRequestCompleted = null;
+	}
+
+	startDataRequest({ orgUnitIds, userSettings }) {
```
Comment (bemailloux) [inline model/dataApiClient.js:24]: should this be async?

**Review by bemailloux — COMMENTED** [review]
Haven't looked in too much detail but makes sense to me. I like how it mirrors the back-end.

File: model/dataApiClient.js (hunk 1 above)
Comment (johngwilkinson) [reply model/dataApiClient.js:24]: Idea is that this creates a promise so all call to fetchResultSet can then await on that same promise. This method itself can return right away though.

**Comment by johngwilkinson** [comment]
I think doing most testing via vis diffs is probably best for simplicity and completeness, but if there's anything that seems to call for unit tests here, please call it out.

File: model/data.js [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+
+export class Data {
+	constructor({ resultSetNames, dataApiClient }) {
+		this._resultSets = resultSetNames.map(name => new ResultSet({ name, dataApiClient }));
+		this._dataApiClient = dataApiClient;
+	}
+
+	load({ orgUnitIds, userSettings }) {
+		this._dataApiClient.startDataRequest({ orgUnitIds, userSettings });
+		this._resultSets.forEach(resultSet => resultSet.load());
```
Comment (Vitalii-Misechko) [inline model/data.js:11]: Should it return an array of promises so a caller can wait on results? It seems we're not going to wait on results but let MobX do the trick

File: route-loader.js [hunk 3]
```diff
@@ -27,10 +27,9 @@ registerRoutes([
 		pattern: '/',
 		view: (ctx) => html`<d2l-insights-adoption-dashboard-view
 							?demo=${ctx.options.isDemo}
```
Comment (Vitalii-Misechko) [inline route-loader.js:29]: we don't need this line right?

**Review by Vitalii-Misechko — COMMENTED** [review]
It seems that Course Access Count doesn't has loading spinner
image

File: demo/ad-demo-view.js [hunk 4]
```diff
@@ -9,6 +9,12 @@ class AdDemoView extends LitElement {
 		// code related to visual diff tests
 		const hashMap = parseHash(window.location.hash);
 
+		const serviceWorker = hashMap.has('immediate') ? worker : waitingWorker;
+		serviceWorker.start();
+
+		// may be used by cards to hard-code things like axis ticks for consistency of visual diffs
+		document.documentElement.setAttribute('isDemo', true);
```
[the remark this answers is not in the document]
Comment (johngwilkinson) [reply demo/ad-demo-view.js:16]: Maybe. This solution comes from engagement, where it was how we fixed a flaky visual diff test. I'm trusting that we ruled out other fixes then.

File: model/data.js (hunk 2 above)
Comment (johngwilkinson) [reply model/data.js:11]: Right, the idea is to let mobx do it. Basically we have data sources which are being monitored by their rendering code: update the data source, and the ui redraws it.

**Comment by johngwilkinson** [comment]
> It seems that Course Access Count doesn't has loading spinner image

Right, it just has stub data and doesn't load from the server yet (mocked or real) - so no spinner is expected yet.

**Review by Vitalii-Misechko — APPROVED** [review]
LGTM!

File: components/cards/system-access.js [hunk 5]
```diff
@@ -237,19 +206,24 @@ class SystemAccessCard extends SkeletonMixin(Localizer(MobxLitElement)) {
...
+			marker: {
 				enabled: data && data.length === 1
 			},
 			color: 'var(--d2l-color-celestine)'
 		}];
 	}
+
+	get data() {
+		// possibly the server will return timestamps and not be nested, in which case this can be simplified
+		return this.resultSet.data.systemAccessData?.map(([d, v]) => [new Date(d).getTime(), v]) || [];
```
Comment (bemailloux) [inline components/cards/system-access.js:222]: Is it ok for this value to be `undefined`? Or does it have to be an array?

**Review by bemailloux — APPROVED** [review]
LGTM!

**Pull request merged**